designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,153 @@
+avoided.
+
+## Detailed Design
+
+The proposed implementation is to modify the `--quiet` flag so that ESLint will skip
+running any rules that are set to the `warn` level.
+
+From an API perspective, this would be implemented by a filter function that filters down to
+which rules should be run. The function would take a list of the rule configuration `(ruleId, severity, rule)`,
+and return the rule that should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+In `cli.js`'s `translateOptions` function, the `filterRules` option should be assigned to
+a function that filters to rules with a `severity` of 2 (error) when the `--quiet` flag is applied, otherwise
+acts as an identity function. In `eslint/flat-eslint.js`, the `filterRules` should be taken from
+the `eslintOptions` object, and passed down to the `linter.verifyAndFix` call.
+
+Within `linter.js`, the API should be added to `VerifyOptions`, and will be passed down into and
+utilized within the `runRules` function before the current `configuredRules` loop. Rather than the
+current configuredRules loop, this should extract the `severity` and `rule` existence checks and
+build a list of `{ ruleId, severity, rule}` objects. This new list should be passed to `filterRules`,
+and the resulting list should be iterated on instead of `configuredRules`.
+
+`normalizeVerifyOptions` should verify that the `filterRules` option is a function, and replace it
+with an identity function if not. `processOptions` in `eslint-helpers.js` should also perform a
+validation check that the `filterRules` option is a function.
+
+The new `filterRules` function when implemented would look like this, using the `--quiet` flag
+rules outlined in this RFC as an example:
+
+```typescript
+linter.verifyAndFix(text, configs, {
+  filterRules: (rules: { ruleId: string; rule: Rule; severity: number }[]) => {
+    return rules.filter((rule) => rule.severity === 2);
+  },
+});
+```
+
+The function acts as a filter on the rules to enable, where removal from the list disables a rule.
+This function would return all rules which have a severity of 2 (error), effectively filtering out
+all other rules.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--quiet`. As the
+rules are not actually run, an assumption would have to be made that all directives
+on rules present only in the un-filtered list are used when in this mode. This is a reasonable
+assumption, as the user likely does not expect `warn` flags to be touched at all in this mode.
+This would also apply to blanket `eslint-disable` directives that disable all rules, which
+should always be assumed to be used while a non-identity filter function is passed.
```

Comment:
I'm happy to switch to length. My original thought around an identity check is it would still trigger on cases where a filter was in use but didn't match anything, but that's realistically not useful at all. I can just add a "isFiltered" boolean check at the start based on length and go from there

Reply:
Sounds good to me. If the filter function, whether it's an identity function or function that filters out warnings or a user-provided function, didn't really filter out any rules, then we're running the full set of configured rules for that file, so blanket `eslint-disable` directives are expected to have an effect in that run. If a directive is not suppressing any problems from the full set of rules, we should mark it as unused.